birdhouse/config/canarie-api/docker_configuration.py.template

```diff
@@ -12,11 +18,97 @@ DATABASE = {
     'log_pid': '/var/run/nginx.pid'
 }
 
+
+# Below is to avoid rate-limiting as much as possible on multiple runs.
+# Because the logparser and monitoring must load the application context and configuration,
+# the requests performed to retrieve version numbers and release times occurs on every cron interval.
+REQUEST_CACHE = "/tmp/canarieapi-config-cache"
+os.makedirs(REQUEST_CACHE, exist_ok=True)
+SESSION = requests_cache.CachedSession(
+    "canarieapi-config-cache",
+    use_temp=True,
+    backend="filesystem",
+    expire_after=datetime.timedelta(days=1),
+    cache_control=False,  # prefer our request cache options, not ones provided by responses
+    stale_if_error=True,
+    allowable_codes=[200],
+    allowable_methods=["GET"],
+)
+
+
+def _get_release_time_from_github_tag(repository_reference: str, tagged_version: str) -> str:
+    repo = repository_reference.split("github.com/")[-1]
+    url = f"https://api.github.com/repos/{repo}/tags"
+    resp = SESSION.get(url, headers={"Accept": "application/json"})
+    if resp.status_code != 200:
+        raise ValueError(
+            f"HTTP status from tags list: {resp.status_code} from [{url}] with detail:\n{resp.text!s}"
+        )
+
+    for tag_info in resp.json():
+        tag_ver = tag_info.get("name")
+        tag_commit_url = tag_info.get("commit", {}).get("url")
+        if tag_ver == tagged_version or tag_ver == "v" + tagged_version and tag_commit_url:
+            break
+    else:
+        raise ValueError(f"No matching version in tagged releases under [{url}].")
+
+    resp = SESSION.get(tag_commit_url, headers={"Accept": "application/json"})
+    if resp.status_code != 200:
+        raise ValueError(
+            f"HTTP status from tag info: {resp.status_code} from [{tag_commit_url}] with detail:\n{resp.text!s}"
+        )
+
+    commit = resp.json()["commit"]
+    commit_date = commit["committer"]["date"]
+    return commit_date
+
+
+def _get_release_time_from_docker_tag(repository_reference: str, tagged_version: str) -> str:
+    ns, repo = repository_reference.split("/", 1)
+    url = f"https://hub.docker.com/v2/namespaces/{ns}/repositories/{repo}/tags/{tagged_version}"
+    resp = SESSION.get(url, headers={"Accept": "application/json"})
+    if resp.status_code != 200:
+        raise ValueError(
+            f"HTTP status from tag info: {resp.status_code} from [{url}] with detail:\n{resp.text!s}"
+        )
+
+    updated = resp.json()["last_updated"]
+    return updated
+
+
+def get_release_time_from_repo_tag(repository_type: str, repository_reference: str, tagged_version: str) -> str:
```

Comment:
Since the links for the "source", "releasenotes", etc. point more often at the GitHub repository, I felt the "releaseTime" was better aligned with it if using the same reference. I'm not against adjusting them for their corresponding docker tag if those feel more adequate. 

I used Docker only for cases where there was a strong divergence in the docker tag vs source tag, such as in the case of GeoServer.

Reply:
Ah ok, I was just curious.  Now that you already invested into github releasetime, and dealt with github rate limiting, let's keep it.

Reply:
Maybe this is a weird question but is there a reason why we need to report these release times by querying the github or dockerhub urls? If we give the user a link to the source and a version number they should be able to go to that website and see the release time for themselves.

I guess that including the release time was a requirement of the canarie folks? Other than that, what is the use-case for displaying the release time of specific components through canarie-api?

Reply:
The fields that are returned were specifically required by Canarie with a specific format.
It is also good to have the release time be displayed at a glance for all components, rather than manually checking them one by one digging down Github links.
Updating them with this PR, I've also seen that almost every component was not reporting the correct version number and release time of what was actually deployed. This approach with requesting the tag release times removes the burden of remembering to update this information each time we update a service.